### ISSUE TITLE
Problem sending large messages (10MB very slow, 100MB never pass)

### ISSUE THREAD

**Issue by 5tan** [emqx/emqx#4787]
**Environment**:

- EMQ X version (e.g. `emqx_ctl status`): 

[4.3.0](https://hub.docker.com/layers/emqx/emqx/4.3.0/images/sha256-487dea78e6fd28fe67fe088ad65043562ebe0a7023a8b0845a8f6de0cad4023f?context=explore)

- Hardware configuration (e.g. `lscpu`):

```$ lscpu 
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
CPU(s):                          16
On-line CPU(s) list:             0-15
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       AuthenticAMD
CPU family:                      23
Model:                           96
Model name:                      AMD Ryzen 7 4800H with Radeon Graphics
Stepping:                        1
Frequency boost:                 enabled
CPU MHz:                         1397.197
CPU max MHz:                     2900,0000
CPU min MHz:                     1400,0000
BogoMIPS:                        5788.91
Virtualization:                  AMD-V
L1d cache:                       256 KiB
L1i cache:                       256 KiB
L2 cache:                        4 MiB
L3 cache:                        8 MiB
NUMA node0 CPU(s):               0-15
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Full AMD retpoline, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht 
                                 syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid
                                  aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand
                                  lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tc
                                 e topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs 
                                 ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsav
                                 eopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr
                                  wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter
                                  pfthreshold avic v_vmsave_vmload vgif umip rdpid overflow_recov succor smca

```

- OS (e.g. `cat /etc/os-release`):

```
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.2 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.2 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

- Kernel (e.g. `uname -a`):

```
$ uname -a
Linux legion 5.4.0-72-generic #80-Ubuntu SMP Mon Apr 12 17:35:00 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

- Erlang/OTP version (in case you build emqx from source code): 

N/A (using emqx docker)

- Others:

  * Docker version 19.03.12, build 48a66213fe
  * client: MQTT.js v3.0.0 (but I don't think it matters)
  * node v14.16.1

**What happened and what you expected to happen**:

I am unable to send/receive large messages using emqx. If I try to send 10MB message it takes about 10 seconds to get it received on localhost. I was unable to send 100MB message. Client (MQTT.js) retries, but gives no error message. 

There are no instant errors in EMQX console/logs, but when I left client retrying for some long time (30 minutes?), I noticed this in logs:
```
2021-05-12T11:36:22.436351+00:00 [warning] [Alarm Handler] Alarm high_system_memory_usage is activated, System memory usage is higher than 70%
2021-05-12T11:38:55.012298+00:00 [error] mqttjs_9b3ac9de@172.19.0.1:53916 [MQTT] Socket error: einval
```

**How to reproduce it (as minimally and precisely as possible)**:

Try to send & receive 100MB message using any client. Below you will find two examples: using MQTT.js client & using BASH/mosquitto-clients.

### MQTT.js example

1. Run `node ./receiver.js`:
```
var mqtt = require('mqtt')
var client  = mqtt.connect('mqtt://localhost', {port: 1883})
 
client.on('connect', function () {
  client.subscribe('test', function (err) {
    if (err) {
      console.log("error: " + err)
    }
  })
})
 
client.on('message', function (topic, message) {
  console.log(message.length)
}
```
2. Run `node ./sender.js`:
```
var mqtt = require('mqtt')
var client  = mqtt.connect('mqtt://localhost', {port: 1883})

buff = new Buffer.alloc(100000000, 's')
console.log("generated")

client.on('error', function (err) {
  console.log("error", err)
})

client.stream.on('error', (err) => {
  console.log('stream error', err);
  client.end()
});

client.on('connect', function () {
  console.log("connected")
  client.publish('test', buff)
  console.log("sent")
})
```

### BASH / mosquitto-clients example

1. Start subscribing client: `mosquitto_sub -h localhost -p 1883 -t test`
2. Publish 100MB message: `perl -e 'print "X" x 100000000' | mosquitto_pub -h localhost -p 1883 -t test -l`

**Anything else we need to know?**:

I have set `mqtt.max_packet_size = 268435455` in `emqx.local.conf`.

**MQTT.js example from above works perfect with mosquitto broker!** "BASH/mosquitto-clients" also works (but slow).

**Comment by zmstone** [comment]
@5tan will fix.

**Comment by zmstone** [comment]
@5tan we just released 4.3.1 check it out. 100MB file send in my local test takes less than 2 seconds.

**Comment by 5tan** [comment]
wow, that was fast! Thank you!
